sdks/python/apache_beam/io/gcp/datastore/v1new/datastoreio.py

```diff
@@ -340,12 +340,13 @@ def finish_bundle(self):
     def _init_batch(self):
       self._batch_bytes_size = 0
       self._batch = self._client.batch()
-      self._batch.begin()
+      self._batch_mutations = []
 
     def _flush_batch(self):
       # Flush the current batch of mutations to Cloud Datastore.
       latency_ms = helper.write_mutations(
```

Comment:
Having both _batch_mutations and _batch seems redundant.
I would move `write_mutations` into this class and create the batch object in it.
You could then take advantage of `add_element_to_batch` to populate the batch.

Reply:
Regarding ByteSize(), we could split add_element_to_batch into 2 part (just a suggestion):

```py
    def element_to_client_batch_item(self, element):
      if not isinstance(element, types.Entity):
        raise ValueError('apache_beam.io.gcp.datastore.v1new.datastoreio.Entity'
                         ' expected, got: %s' % type(element))
      if not element.key.project:
        element.key.project = self._project
      client_entity = element.to_client_entity()
      if client_entity.key.is_partial:
        raise ValueError('Entities to be written to Cloud Datastore must '
                         'have complete keys:\n%s' % client_entity)
      return client_entity

    def add_to_batch(self, client_batch_item):
      self._batch.put(client_entity)
```

Reply:
I've made related changes but still couldn't avoid storing batch and raw batch elements separately.
We could skip the rest of elements (over the bytes limit) when building the batch in `write_mutations` and let them be picked up on the next batch (so we won't purge _batch but there might be some elements left from previous batch) -- but it seems to create more confusion.

Reply:
I think your choice is a valid compromise with what to do with the results of `element_to_client_batch_item`. I see 2 options here: 
1. Save the batch items.
2. Discard batch items after calling `ByteSize()` on them.

The first option seems more CPU efficient while the second seems more memory efficient. I don't know which option is faster, but from my limited experience saving CPU at the expense of RAM seems like a good tradeoff.